providers/implementations/ciphers/cipher_chacha20.c

```diff
@@ -64,6 +65,24 @@ static void chacha20_freectx(void *vctx)
     }
 }
 
+static void *chacha20_dupctx(void *vctx)
+{
+    PROV_CHACHA20_CTX *ctx = (PROV_CHACHA20_CTX *)vctx;
+    PROV_CHACHA20_CTX *dupctx = NULL;
+
+    if (ctx != NULL) {
+        dupctx = OPENSSL_memdup(ctx, sizeof(*dupctx));
+        if (dupctx != NULL && dupctx->base.tlsmac != NULL) {
+            dupctx->base.tlsmac = OPENSSL_memdup(dupctx->base.tlsmac, dupctx->base.tlsmacsize);        
+            if (dupctx->base.tlsmac == NULL) {
+                OPENSSL_free(dupctx);
+                dupctx = NULL;
+            }
+        }
```

Comment:
I was looking at how the tlsmac is used & set in the code and it is unfortunately a little bit more complicated. It is sometimes allocated and sometimes it is a pointer inside the output buffer - that is super ugly and undocumented but I assume it works for the TLS use-case.

So there are two options:
1. Do the memdup of tlsmac only if dupctx->base.alloced is set, otherwise just copy the pointer which is implicitly done by the memdup above.
2. Just nullify the tlsmac on dupctx().

BTW none of the other dupctx functions handle the tlsmac properly as they just copy the context and do not bother with neither 1 nor 2.

Given the above and the super-ugly behavior of tlsmac in case of stream ciphers I would be inclined to do 2.

Reply:
ACK, I'll make the change.  I've also added an issue to the project board to audit the other ciphers

Reply:
fixed